codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/waiters/KotlinJmespathExpressionVisitor.kt

```diff
@@ -57,10 +57,14 @@ class KotlinJmespathExpressionVisitor(
     private val currentShape: Shape
         get() = shapeCursor.last()
 
+    // traverses an independent expression (one whose resolved scope does not persist in the outer evaluation)
     private fun acceptSubexpression(expr: JmespathExpression): VisitedExpression {
-        shapeCursor.addLast(currentShape)
+        val pos = shapeCursor.size
         val out = expr.accept(this)
-        shapeCursor.removeLast()
+
+        val diff = shapeCursor.size - pos
+        repeat(diff) { shapeCursor.removeLast() } // reset the shape cursor
```

Comment:
**Question**: Are some of our `visit*` methods putting something on the stack and failing to remove it at the end? Should we be wrapping stack pushes such that they're always definitively popped at the end of a scope?

```kotlin
override fun visitFoo(expr: FooExpression): VisitedExpression {
    val shape = shapeForFoo(expr)
    return shapeCursor.withPush(shape) { // push
        // do some things for foo codegen
        VisitedExpression(foo, bar)
    } // pop
}
```

Reply:
The visitors aren't really written to be "self-popping" per se, inner expressions are generally opaque to us, so when you're visiting some expression, you don't always have context on whether you're safe to pop back after visiting some given sub-expression. Some binary expressions will have sub-tree on the left where we very much need the modified scope to stay in context when we go to the right (i.e. projections).

What we _do_ know without tree introspection is when some sub-tree has isolated scope, so we can jump into it, let it do its business, and then rewind.

I think it's feasible to have implemented it in the way you're describing, though I think there'd have been more code to have expressions pop their own state off when done based on surrounding context.

Reply:
OK, we can leave it as is for now.